File: app/src/main/java/com/merxury/ui/ComponentActivity.java
package com.merxury.ui;

import android.animation.ArgbEvaluator;
import android.animation.ValueAnimator;
import android.annotation.SuppressLint;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.drawable.ColorDrawable;
import android.os.Bundle;
import android.support.constraint.ConstraintLayout;
import android.support.design.widget.AppBarLayout;
import android.support.design.widget.CollapsingToolbarLayout;
import android.support.design.widget.TabLayout;
import android.support.v4.content.ContextCompat;
import android.support.v4.view.ViewPager;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions;
import com.merxury.adapter.FragmentAdapter;
import com.merxury.blocker.R;
import com.merxury.constant.Constant;
import com.merxury.entity.Application;
import com.merxury.fragment.ComponentFragment;
import com.merxury.utils.ApplicationUtils;
import com.mikepenz.materialdrawer.Drawer;
import com.mikepenz.materialdrawer.DrawerBuilder;

import butterknife.BindView;
import butterknife.ButterKnife;

public class ComponentActivity extends AppCompatActivity {

    @BindView(R.id.component_toolbar)
    Toolbar mToolbar;
    @BindView(R.id.component_appbar)
    AppBarLayout mAppBar;
    @BindView(R.id.component_collapsing_toolbar)
    CollapsingToolbarLayout mCollapsingToolbarLayout;
    @BindView(R.id.component_tabs)
    TabLayout mTabLayout;
    @BindView(R.id.component_viewpager)
    ViewPager mViewPager;
    @BindView(R.id.app_info_layout)
    ConstraintLayout mAppBriefLayout;
    @BindView(R.id.app_info_app_name)
    TextView mAppName;
    @BindView(R.id.app_info_icon)
    ImageView mAppIcon;
    @BindView(R.id.app_info_app_package_name)
    TextView mAppPackageName;
    @BindView(R.id.app_info_target_sdk_version)
    TextView mTargetSdkVersion;
    @BindView(R.id.app_info_min_sdk_version)
    TextView mMinSdkVersion;

    private Application mApplicationInfo;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_component);
        ButterKnife.bind(this);
        Intent intent = getIntent();
        mApplicationInfo = intent.getParcelableExtra(Constant.APPLICATION);
        initActionBar();
        initAppBriefInfoLayout();
        initDrawer();
        initViewPager();
        initTab();
        initOnClickListener();
    }

    @Override
    protected void onStop() {
        super.onStop();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
    }

    @Override
    protected void onResume() {
        super.onResume();
    }

    private void initDrawer() {
        Drawer drawer = new DrawerBuilder().withActivity(this).build();
    }

    private void initViewPager() {
        PackageManager pm = getPackageManager();
        String packageName = mApplicationInfo.getPackageName();
        FragmentAdapter adapter = new FragmentAdapter(getSupportFragmentManager());
        adapter.addFragment(ComponentFragment.getInstance(pm, packageName, ComponentFragment.RECEIVER), getString(R.string.receiver));
        adapter.addFragment(ComponentFragment.getInstance(pm, packageName, ComponentFragment.SERVICE), getString(R.string.service));
        adapter.addFragment(ComponentFragment.getInstance(pm, packageName, ComponentFragment.ACTIVITY), getString(R.string.activity));
        adapter.addFragment(ComponentFragment.getInstance(pm, packageName, ComponentFragment.PROVIDER), getString(R.string.provider));
        mViewPager.setAdapter(adapter);
    }

    private void initTab() {
        mTabLayout.setupWithViewPager(mViewPager);
        changeColor(getColorForTab(0));
        mTabLayout.setSelectedTabIndicatorColor(getResources().getColor(R.color.md_white_1000));
        mTabLayout.addOnTabSelectedListener(new TabLayout.OnTabSelectedListener() {
            @Override
            public void onTabSelected(TabLayout.Tab tab) {
                changeTabBackgroundColor(tab);
            }

            @Override
            public void onTabUnselected(TabLayout.Tab tab) {

            }

            @Override
            public void onTabReselected(TabLayout.Tab tab) {

            }
        });
    }

    private void changeTabBackgroundColor(TabLayout.Tab tab) {
        int colorFrom;
        if (mTabLayout.getBackground() != null) {
            colorFrom = ((ColorDrawable) mTabLayout.getBackground()).getColor();
        } else {
            colorFrom = ContextCompat.getColor(this, android.R.color.darker_gray);
        }
        int colorTo = getColorForTab(tab.getPosition());
        ValueAnimator colorAnimation = ValueAnimator.ofObject(new ArgbEvaluator(), colorFrom, colorTo);
        colorAnimation.addUpdateListener(animation -> {
            int color = (int) animation.getAnimatedValue();
            changeColor(color);
        });
        colorAnimation.setDuration(500);
        colorAnimation.start();
    }

    private int getColorForTab(int position) {
        switch (position) {
            case 0:
                return ContextCompat.getColor(this, R.color.md_blue_700);
            case 1:
                return ContextCompat.getColor(this, R.color.md_light_green_700);
            case 2:
                return ContextCompat.getColor(this, R.color.md_orange_700);
            case 3:
                return ContextCompat.getColor(this, R.color.md_red_700);
            default:
                return ContextCompat.getColor(this, R.color.md_grey_700);
        }
    }

    private void changeColor(int color) {
        mToolbar.setBackgroundColor(color);
        mTabLayout.setBackgroundColor(color);
        mCollapsingToolbarLayout.setBackgroundColor(color);
        getWindow().setStatusBarColor(color);
    }

    private void initActionBar() {
        setSupportActionBar(mToolbar);
        final ActionBar ab = getSupportActionBar();
        if (ab != null) {
            ab.setHomeAsUpIndicator(R.drawable.ic_menu_white_24dp);
            ab.setDisplayHomeAsUpEnabled(true);
            ab.setDisplayShowTitleEnabled(false);
        }
    }

    @SuppressLint("CheckResult")
    private void initAppBriefInfoLayout() {
        mAppName.setText(getString(R.string.application_label, mApplicationInfo.getLabel()));
        mAppPackageName.setText(getString(R.string.package_name, mApplicationInfo.getPackageName()));
        mTargetSdkVersion.setText(getString(R.string.target_sdk_version, mApplicationInfo.getTargetSdkVersion()));
        mMinSdkVersion.setText(getString(R.string.min_sdk_version, mApplicationInfo.getMinSdkVersion()));
        Glide.with(this)
                .load(mApplicationInfo.getApplicationIcon(getPackageManager()))
                .transition(new DrawableTransitionOptions().crossFade())
                .into(mAppIcon);
    }

    private void initOnClickListener() {
        mAppIcon.setOnClickListener((view) -> ApplicationUtils.Companion.startApplication(this, mApplicationInfo.getPackageName()));
    }
}

File: app/src/main/java/com/merxury/blocker/ui/component/ComponentActivity.kt
package com.merxury.blocker.ui.component

import android.animation.ArgbEvaluator
import android.animation.ValueAnimator
import android.graphics.drawable.ColorDrawable
import android.os.Bundle
import android.util.SparseArray
import android.view.MenuItem
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.SearchView
import androidx.core.content.ContextCompat
import com.google.android.material.tabs.TabLayout
import com.merxury.blocker.R
import com.merxury.blocker.adapter.FragmentAdapter
import com.merxury.blocker.base.IActivityView
import com.merxury.blocker.ui.Constants
import com.merxury.blocker.util.AppLauncher
import com.merxury.blocker.util.setupActionBar
import com.merxury.libkit.entity.Application
import com.merxury.libkit.utils.StatusBarUtil
import kotlinx.android.synthetic.main.activity_component.*
import kotlinx.android.synthetic.main.application_brief_info_layout.*
import kotlinx.coroutines.*
import java.lang.RuntimeException
import kotlin.coroutines.CoroutineContext

class ComponentActivity : AppCompatActivity(), IActivityView, CoroutineScope {
    private lateinit var application: Application
    private lateinit var adapter: FragmentAdapter
    private lateinit var job: Job
    override val coroutineContext: CoroutineContext
        get() = Dispatchers.Main + job

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        job = Job()
        setContentView(R.layout.activity_component)
        setupActionBar(R.id.component_toolbar) {
            setDisplayHomeAsUpEnabled(true)
            setDisplayShowTitleEnabled(false)
        }
        getDataFromIntent()
        setupViewPager()
        setupTab()
        showApplicationBriefInfo(application)
    }

    override fun onDestroy() {
        super.onDestroy()
        job.cancel()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> finish()
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onBackPressed() {
        findViewById<SearchView>(R.id.menu_search)?.let {
            if (!it.isIconified) {
                it.isIconified = true
                it.clearFocus()
                return
            }
        }
        super.onBackPressed()
    }

    override fun getBackgroundColor(tabPosition: Int): Int {
        return when (tabPosition) {
            0 -> ContextCompat.getColor(this, R.color.google_blue)
            1 -> ContextCompat.getColor(this, R.color.google_green)
            2 -> ContextCompat.getColor(this, R.color.google_red)
            3 -> ContextCompat.getColor(this, R.color.google_yellow)
            else -> ContextCompat.getColor(this, R.color.secondary_text)
        }
    }

    private fun setupViewPager() {
        adapter = FragmentAdapter(supportFragmentManager)
        adapter.addFragment(ComponentFragment.newInstance(application.packageName, EComponentType.RECEIVER), getString(R.string.receiver))
        adapter.addFragment(ComponentFragment.newInstance(application.packageName, EComponentType.SERVICE), getString(R.string.service))
        adapter.addFragment(ComponentFragment.newInstance(application.packageName, EComponentType.ACTIVITY), getString(R.string.activity))
        adapter.addFragment(ComponentFragment.newInstance(application.packageName, EComponentType.PROVIDER), getString(R.string.provider))
        component_viewpager.offscreenPageLimit = 3
        component_viewpager.adapter = adapter
    }

    private fun getDataFromIntent() {
        if (intent == null) {
            finish()
        }
        application = intent.getParcelableExtra(Constants.APPLICATION) ?:
                throw RuntimeException("App info should not be null")
    }

    private fun setupTab() {
        component_tabs.setupWithViewPager(component_viewpager)
        changeColor(getBackgroundColor(0))
        component_tabs.setSelectedTabIndicatorColor(ContextCompat.getColor(this, R.color.info_text_color))
        component_tabs.addOnTabSelectedListener(object : TabLayout.OnTabSelectedListener {
            override fun onTabSelected(tab: TabLayout.Tab) {
                changeBackgroundColor(tab)
            }

            override fun onTabUnselected(tab: TabLayout.Tab) {

            }

            override fun onTabReselected(tab: TabLayout.Tab) {

            }
        })
    }

    private fun showApplicationBriefInfo(application: Application) {
        app_info_app_name.text = getString(R.string.application_label, application.label)
        app_info_app_package_name.text = getString(R.string.package_name, application.packageName)
        app_info_target_sdk_version.text = getString(R.string.target_sdk_version, CODENAME.get(application.targetSdkVersion, UNKNOWN))
        app_info_min_sdk_version.text = getString(R.string.min_sdk_version, CODENAME.get(application.minSdkVersion, UNKNOWN))
        launch {
            val icon = withContext(Dispatchers.IO) {
                application.getApplicationIcon(packageManager)
            }
            app_info_icon.setImageDrawable(icon)
        }
        app_info_icon.setOnClickListener { AppLauncher.startApplication(this, application.packageName) }
    }

    private fun changeColor(color: Int) {
        component_toolbar.setBackgroundColor(color)
        component_tabs.setBackgroundColor(color)
        component_collapsing_toolbar.setBackgroundColor(color)
        StatusBarUtil.setColor(this, color, com.merxury.blocker.constant.Constant.STATUS_BAR_ALPHA)
    }

    private fun changeBackgroundColor(tab: TabLayout.Tab) {
        val colorFrom = if (component_tabs.background != null) {
            (component_tabs.background as ColorDrawable).color
        } else {
            ContextCompat.getColor(this, android.R.color.darker_gray)
        }
        val colorTo = getBackgroundColor(tab.position)
        val colorAnimation = ValueAnimator.ofObject(ArgbEvaluator(), colorFrom, colorTo)
        colorAnimation.addUpdateListener { animation ->
            val color = animation.animatedValue as Int
            changeColor(color)
        }
        colorAnimation.duration = 500
        colorAnimation.start()
    }

    companion object {
        private val CODENAME: SparseArray<String> = SparseArray(32)
        private const val UNKNOWN = "Unknown"

        init {
            CODENAME.put(1, "Android 1.0")
            CODENAME.put(2, "Android 1.1")
            CODENAME.put(3, "Android 1.5")
            CODENAME.put(4, "Android 1.6")
            CODENAME.put(5, "Android 2.0")
            CODENAME.put(6, "Android 2.0.1")
            CODENAME.put(7, "Android 2.1")
            CODENAME.put(8, "Android 2.2")
            CODENAME.put(9, "Android 2.3")
            CODENAME.put(10, "Android 2.3.3")
            CODENAME.put(11, "Android 3.0")
            CODENAME.put(12, "Android 3.1")
            CODENAME.put(13, "Android 3.2")
            CODENAME.put(14, "Android 4.0.1")
            CODENAME.put(15, "Android 4.0.3")
            CODENAME.put(16, "Android 4.1")
            CODENAME.put(17, "Android 4.2")
            CODENAME.put(18, "Android 4.3")
            CODENAME.put(19, "Android 4.4")
            CODENAME.put(21, "Android 5.0")
            CODENAME.put(22, "Android 5.1")
            CODENAME.put(23, "Android 6.0")
            CODENAME.put(24, "Android 7.0")
            CODENAME.put(25, "Android 7.1")
            CODENAME.put(26, "Android 8.0")
            CODENAME.put(27, "Android 8.1")
            CODENAME.put(28, "Android P")
            CODENAME.put(29, "Android 10")
            CODENAME.put(30, "Android R")
            CODENAME.put(31, "Android S")
            CODENAME.put(32, "Android T")
            CODENAME.put(33, "Android U")
            // Reference : https://source.android.com/setup/start/build-numbers
        }
    }
}
